csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -182,11 +182,239 @@ bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+      ref_broadcast_mask.reserve(root.size());
+      for (const auto i : c10::irange(root.size())) {
+        ref_broadcast_mask.push_back(!root.at(i)->isReduction());
+      }
+      break;
+    }
+  }
+  NVF_ERROR(!ref_broadcast_mask.empty(), "ref_broadcast_mask is empty!");
+
+  // find the brodcast tensor whose broadcast makse is same to the reference
+  std::vector<TensorView*> outer_broadcast_tvs;
+  for (auto tv : ir_utils::allTvs(fusion)) {
+    if (std::any_of(
+            tv->getLeafDomain().begin(),
+            tv->getLeafDomain().end(),
+            [](IterDomain* id) { return id->isBroadcast(); })) {
+      if (auto bcast = dynamic_cast<BroadcastOp*>(tv->definition())) {
+        if (bcast->getBroadcastDimFlags() == ref_broadcast_mask) {
+          outer_broadcast_tvs.emplace_back(tv);
+        }
+      }
+    }
+  }
+  return outer_broadcast_tvs;
+}
+
+// Size of buffers storing intermediate outer reduction results
+int64_t partialOuterReductionBufferSize(
+    const std::vector<TensorView*>& reduction_tvs,
+    SchedulerRuntimeInfo& runtime_info) {
+  int64_t partial_reduction_buffer_size = 0;
+  for (auto buffer : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(buffer)) {
+      continue;
+    }
+    int64_t buffer_size = -1;
+    for (auto id : buffer->getMaybeRFactorDomain()) {
+      if (id->isReduction() || id->isBroadcast()) {
+        continue;
+      }
+      auto id_size = runtime_info.expressionEvaluator().evaluate(id->extent());
+      NVF_ERROR(id_size.hasValue(), "Could not infer persistent buffer size.");
+      if (buffer_size == -1) {
+        buffer_size = id_size.as<int64_t>();
+      } else {
+        buffer_size *= id_size.as<int64_t>();
+      }
+    }
+    buffer_size = (buffer_size == -1) ? 0
+                                      : buffer_size *
+            (int64_t)dataTypeSize(buffer->getDataType().value(),
+                                  runtime_info.getIndexType());
+    partial_reduction_buffer_size += buffer_size;
+  }
+  return partial_reduction_buffer_size;
+}
+
+// Sorts projectable_buffer_inputs based on the latency from smallest to
+// largest. The latency is calculated based on the number of uses, the buffer
+// data type size, and the pre-defined schedule structures. Assume each
+// global memory load has latency 100, each shared memory load has latency 10,
+// latency is proportional to the data type size. There is no write from
+// register to shared memory, so this type of latency is not considered. The
+// magic number 100 and 10 are not critical for our current major use case, e.g.
+// layer norm backward, and rms norm backward. The outer broadcast tensor (the
+// weight tensor) is almost always moved to shared memory before other buffers
+// unless the outer dimension is very small, e.g. less than 1 sm count or other
+// buffers are used 10 times more than the outer broadcast tensor.
+std::vector<TensorView*> sortPersistentBuffers(
```

Comment:
nitpick: the function is called `sortPersistentBuffers` and there's an argument of `persistent_buffers`. But it's instead sorting `projectable_buffe_inputs`.

Reply:
I agree, it is a little bit confusing. The reason is because the current innerOuter scheduler always project to inputs. I renamed this function and added another one to process cases we won't project to inputs (when there is no out broadcast tv, see the newly added test case)